### PR TITLE
Feat: issue #427 implementation

### PR DESCRIPTION
**Pull request by arkanoider**
Implementation of #427 :
User should just have one waiting-payment or waiting-buyer-invoice order pending when taking orders ( sell or buy ), in case of other pending orders they receive a CantDo message ( maybe we can add a more meaningful one )

### CODE REVIEW

**Review by Catrya — COMMENTED** [review]
Hi @arkanoider Good job. However, I found 2 cases where it can be improved because it shouldn't work like this.

1)
- A creates an order
- B takes A's order but doesn't respond
- A tries to take C's order and Mostro doesn't let him (A doesn't know that his order was taken because the taker doesn't continue showing that he doesn't let A know to do his part)

 This could be used by a malicious user to block other users from taking orders

How it should work is that A could take an order even if it is in status `waiting-buyer-invoice` or `waiting-payment` if it is the taker's turn

2) 
- A takes an order from B and does his part
- B doesn't respond and leaves A waiting
- A tries to take an order from C and Mostro doesn't let him

Mostro should allow A to take another order if he has already done his part

**Comment by arkanoider** [comment]
@Catrya need some more refinement, you're right. Have to check how to manage better all the possible cases.

**Comment by Catrya** [comment]
> > * A creates an order
> > * B takes A's order but doesn't respond
> > * A tries to take C's order and Mostro doesn't let him (A doesn't know that his order was taken because the taker doesn't continue showing that he doesn't let A know to do his part)
> > 
> > This could be used by a malicious user to block other users from taking orders
> > How it should work is that A could take an order even if it is in status `waiting-buyer-invoice` or `waiting-payment` if it is the taker's turn
> 
> So for case 1 @Catrya , just for my mental recap:
> 
> -- A creates a sell order -- B Takes it and then he stops here ( order is `waiting buyer invoice` ) -- A now is blocked to take another order.
> 
> This is the sequence for case 1?

Yes, but Mostro should let A take another order

**Comment by arkanoider** [comment]
> Yes, but Mostro should let A take another order

I tested quickly and with a clean db with two sell orders from A and two from B and B was able to take a new order correctly after A took one. But I could be totally wrong @Catrya please do a round on your side again when you can.

**Comment by Catrya** [comment]
> > Yes, but Mostro should let A take another order
> 
> I tested quickly and with a clean db with two sell orders from A and two from B and B was able to take a new order correctly after A took one. But I could be totally wrong @Catrya please do a round on your side again when you can.

You're right, I just checked again and it's possible.
I don't know what I did wrong yesterday. I'll try again more carefully tomorrow morning.

**Comment by arkanoider** [comment]
> > > Yes, but Mostro should let A take another order
> > 
> > I tested quickly and with a clean db with two sell orders from A and two from B and B was able to take a new order correctly after A took one. But I could be totally wrong @Catrya please do a round on your side again when you can.
> 
> You're right, I just checked again and it's possible.
> I don't know what I did wrong yesterday. I'll try again more carefully tomorrow morning.

maybe some edge cases depending on db situation, do your test meanwhile i will fix event ratings issue.

**Comment by arkanoider** [comment]
Will try to do another round of testing on this tonite or tomorrow morning @Catrya . Let me know if you found wrong behaviour of this PR.

File: src/app/take_buy.rs [hunk 1]
```diff
@@ -25,6 +25,11 @@ pub async fn take_buy_action(
     // Get the request ID from the message
     let request_id = msg.get_inner_message_kind().request_id;
 
+    // Check if the buyer has a pending order
+    if seller_has_pending_order(pool, event.sender.to_string()).await? {
+        return Err(MostroCantDo(CantDoReason::InvalidAction));
+    }
```
[the remark this answers is not in the document]
Comment (grunch) [reply src/app/take_buy.rs:31]: Agree with the rabbit

[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app/take_buy.rs:31]: > Agree with the rabbit

I will add this to mostro core with my fix for #447, will do a feature branch also if maybe this could be also a `hot-fix` case.

**Comment by Catrya** [comment]
> Will try to do another round of testing on this tonite or tomorrow morning @Catrya . Let me know if you found wrong behaviour of this PR.

Hi @arkanoider this case have to be fixed:

- A takes an order from B and does his part
- B doesn't respond and leaves A waiting
- A tries to take an order from C and Mostro doesn't let him
Mostro should allow A to take another order if he has already done his part

I tested again and the first case that I mentioned worked fine as I told you later.
But the second one didn't, that one hasn't been fixed yet

**Comment by arkanoider** [comment]
> > Will try to do another round of testing on this tonite or tomorrow morning @Catrya . Let me know if you found wrong behaviour of this PR.
> 
> Hi @arkanoider this case have to be fixed:
> 
> - A takes an order from B and does his part
> - B doesn't respond and leaves A waiting
> - A tries to take an order from C and Mostro doesn't let him
> Mostro should allow A to take another order if he has already done his part
> 
> I tested again and the first case that I mentioned worked fine as I told you later.
> But the second one didn't, that one hasn't been fixed yet

confirm i did not retest swcond case, will be on it tonite or tomorrow.

File: src/db.rs [hunk 2]
```diff
@@ -448,6 +453,60 @@ pub async fn update_user_trade_index(
...
+            SELECT EXISTS (SELECT 1 FROM orders WHERE master_buyer_pubkey = ?1 AND status = 'waiting-buyer-invoice')
+        "#,
+    )
+    .bind(pubkey)
+    .map(|row: SqliteRow| row.get(0))
+    .fetch_one(&mut conn)
+    .await.map_err(|e| MostroInternalErr(ServiceError::DbAccessError(e.to_string())))?;
+
+    Ok(rows_affected)
+}
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/db.rs]: fixed comments!

**Comment by arkanoider** [comment]
Ok @Catrya @grunch 

i did some tests and seems working with the right logic now, do a round on your side if you see still some edge cases.

**Comment by Catrya** [comment]
> * A takes an order from B and does his part
> * B doesn't respond and leaves A waiting
> * A tries to take an order from C and Mostro doesn't let him
>   Mostro should allow A to take another order if he has already done his part

Hi @arkanoider I still find this error in my side

**Comment by Catrya** [comment]
tACK @grunch  @arkanoider

**Comment by arkanoider** [comment]
Conflicts are due to latest mostro-core in toml. Let me know if I can resolve and merge @grunch ...

**Comment by grunch** [comment]
> Conflicts are due to latest mostro-core in toml. Let me know if I can resolve and merge @grunch ...
> 

please solve those conflicts, we can't merge like this

**Comment by arkanoider** [comment]
> > Conflicts are due to latest mostro-core in toml. Let me know if I can resolve and merge @grunch ...
> 
> please solve those conflicts, we can't merge like this

Done

**Review by grunch — APPROVED** [review]
LGTM
